### PR TITLE
[CELEBORN-1423][HELM] Refactor chart templates and update readme docs

### PR DESCRIPTION
**Pull request by ChenYi015**
### What changes were proposed in this pull request?

- Comment license with helm template syntax
- It's better that all resource names are defined once in `_helpers.tml` and refered in other files.
- Format yaml template files with kubernetes style (specifically, follow the output of `kubectl`) i.e. yaml sequence does not have indentation.
- Refactor service and stetefulset templates.
- Update helm chart values and README file.
- Add `.helmignore` file to specify patterns for files and directories that should be ignored when packaging the chart.

### Why are the changes needed?

As dicussed in #2499.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Local test.

### CODE REVIEW

**Comment by ChenYi015** [comment]
Ping @pan3793 , PTAL, thanks.

**Comment by pan3793** [comment]
Please use a new JIRA ticket, LGTM except for the following items:

> - Add README template file `README.md.gotmpl`, and use [helm-docs](https://github.com/norwoodj/helm-docs)  to generate `README.md` file automatically.
> - Add a pre-commit git hook to run to generate the latest helm chart README.md file automatically.

Celeborn does use a templating approach to render the configurations page, but I would say `value.yaml` is a different case, unlike we define configurations in the Scala code, YAML itself is a format optimized for user reading, and end users are likely to read and modify the `values.yaml`. I tend to embed docs and examples into `value.yaml` as-is.

**Comment by ChenYi015** [comment]
@pan3793 I had opened a new JIRA ticket and udpate the PR title. For `values.yaml`, I had already avoided using confusing annotations like `@ignore`, and only some `--` are added before comments.  I think that would not affect the users' reading experience.

**Comment by ChenYi015** [comment]
Besides, I had added some configuration examples in `values.yaml` like:

```yaml
resources:
  # -- Pod resources for Celeborn master pods
  master: {}
    # limits:
    #   cpu: 100m
    #   memory: 128Mi
    # requests:
    #   cpu: 100m
    #   memory: 128Mi

  # -- Pod resources for Celeborn worker pods
  worker: {}
    # limits:
    #   cpu: 100m
    #   memory: 128Mi
    # requests:
    #   cpu: 100m
    #   memory: 128Mi

# -- Pod annotations
podAnnotations: {}
# key1: value1
# key2: value2

# -- Pod node selector
nodeSelector: {}
# key1: value1
# key2: value2

# -- Pod tolerations
tolerations: []
# - key: key1
#   operator: Equal
#   value: value1
#   effect: NoSchedule
# - key: key2
#   operator: Exists
#   effect: NoSchedule
```

**Comment by pan3793** [comment]
@ChenYi015 The key point is I think the template README introduces complexity(setup and learn helm-docs) but does not gain many benefits.

**Comment by ChenYi015** [comment]
> @ChenYi015 The key point is I think the template README introduces complexity(setup and learn helm-docs) but does not gain many benefits.

@pan3793 One thing I want to do is adding version control to the Helm charts, just like you see in many other Helm repos. This way, whenever we update a chart,  the users can benefit from reading the latest docs. For the complexity of `helm-docs` command, please don't worry, only a `docker run` command is needed, and you do not need to learn how to install and setup helm-docs. I will update the helm developer guide later. Besides, developers do not need to learn go templates either, just update the `README.md.gotmpl` like standard markdown files, content that needs to be generated automatically is already setup.

**Comment by pan3793** [comment]
@ChenYi015 can we separate helm-docs parts and merge other changes first, and discuss it later? 

> ... just like you see in many other Helm repos.

can you give some concrete examples?

**Comment by ChenYi015** [comment]
@pan3793 For example, [Apache Superset](https://github.com/apache/superset/tree/master/helm/superset).

**Comment by ChenYi015** [comment]
@pan3793 I have force-pushed the commits and the helm-docs part was removed.

File: charts/celeborn/README.md [hunk 1]
```diff
@@ -40,17 +41,59 @@ When you want to test the template rendering, but not actually install anything.
...
+More details in [Helm Install](https://helm.sh/docs/helm/helm_install/).
+The chart can be customized using the following [celeborn configurations](https://celeborn.apache.org/docs/latest/configuration/#important-configurations).
+Specify parameters using `--set key=value[,key=value]` argument to `helm install`.
 
 ## Documentation
 
-For additional details on deploying the Celeborn Kubernetes Helm chart, please refer to the [Celeborn on Kubernetes](https://celeborn.apache.org/docs/latest/deploy_on_k8s/) documentation
+For additional details on deploying the Celeborn Kubernetes Helm chart, please refer to the [Celeborn on Kubernetes](https://celeborn.apache.org/docs/latest/deploy_on_k8s/) documentation.
+
+## Values
```
Comment (pan3793) [inline charts/celeborn/README.md:63]: let's remove this table at the current stage

File: charts/celeborn/README.md [hunk 2]
```diff
@@ -19,17 +19,18 @@ under the License.
 
 # Helm Chart for Apache Celeborn
 
-[Apache Celeborn](https://celeborn.apache.org) is an intermediate data service for Big Data compute engines (i.e. ETL, OLAP and Streaming engines) to boost performance, stability, and flexibility. Intermediate data typically include shuffle and spilled data.
+![Version: 0.1.0](https://img.shields.io/badge/Version-0.1.0-informational?style=flat-square) ![Type: application](https://img.shields.io/badge/Type-application-informational?style=flat-square) ![AppVersion: 0.4.0](https://img.shields.io/badge/AppVersion-0.4.0-informational?style=flat-square)
```
Comment (pan3793) [inline charts/celeborn/README.md:22]: remove the badge

File: charts/celeborn/values.yaml [hunk 3]
```diff
@@ -76,10 +87,11 @@ volumes:
       diskType: SSD
       capacity: 100Gi
 
-# celeborn configurations
+# -- Celeborn configurations
 celeborn:
   celeborn.master.ha.enabled: true
   celeborn.metrics.enabled: true
+  celeborn.metrics.prometheus.path: /metrics/prometheus
```
Comment (pan3793) [inline charts/celeborn/values.yaml:94]: a bug fix or something else?

File: charts/celeborn/README.md (hunk 1 above)
Comment (ChenYi015) [reply charts/celeborn/README.md:63]: > let's remove this table at the current stage

@pan3793  Ok, already removed.

File: charts/celeborn/values.yaml (hunk 3 above)
Comment (ChenYi015) [reply charts/celeborn/values.yaml:94]: > a bug fix or something else?

The pod monitor's metric endpoint should respect `celeborn.metrics.prometheus.path`, and I will update in later PRs.

File: charts/celeborn/templates/master/statefulset.yaml [hunk 4]
```diff
@@ -1,134 +1,148 @@
...
-          - "/bin/sh"
-          - '-c'
-          {{- $namespace := .Release.Namespace }}
-          - "until {{ range until (.Values.masterReplicas |int) }}nslookup {{ $.Release.Name }}-master-{{ . }}.{{ $.Release.Name }}-master-svc.{{ $namespace }}.svc.{{ $.Values.cluster.name }}.local && {{ end }}true; do echo waiting for master; sleep 2; done && exec /opt/celeborn/sbin/start-master.sh"
+        - /usr/bin/tini
+        - --
+        - /bin/sh
+        - -c
+        {{- $namespace := .Release.Namespace }}
+        - "until {{ range until (.Values.masterReplicas |int) }}nslookup {{ include "celeborn.masterStatefulSetName" $ }}-{{ . }}.{{ include "celeborn.masterServiceName" $ }}.{{ $namespace }}.svc.{{ $.Values.cluster.name }}.local && {{ end }}true; do echo waiting for master; sleep 2; done && exec /opt/celeborn/sbin/start-master.sh"
```
[the remark this answers is not in the document]
Comment (ChenYi015) [reply charts/celeborn/templates/master/statefulset.yaml]: Updated.

**Comment by ChenYi015** [comment]
@pan3793 I have updated the commits, PTAL, thanks.

File: charts/celeborn/values.yaml (hunk 3 above)
Comment (pan3793) [reply charts/celeborn/values.yaml:94]: we should support the `celeborn.metrics.prometheus.path` absent case too.

**Review by pan3793 — APPROVED** [review]
LGTM, one nit.

**Comment by ChenYi015** [comment]
> LGTM, one nit.

@pan3793 Fixed.

**Comment by pan3793** [comment]
Thanks, merged to main for 0.5

**Pull request closed without merging**